### PR TITLE
Liqun/merge e2e pipelines

### PR DESCRIPTION
**Pull request by liqunfu**
**Description**: Merge cpp and python e2e nightly pipeline.

**Motivation and Context**
improve e2e test

### CODE REVIEW

File: tools/ci_build/build.py [hunk 1]
```diff
@@ -1354,13 +1354,6 @@ def run_onnxruntime_tests(args, source_dir, ctest_path, build_dir, configs):
         log.info("Running tests for %s configuration", config)
         cwd = get_config_build_dir(build_dir, config)
 
-        if args.enable_training and args.use_cuda and args.enable_training_python_frontend_e2e_tests:
```
Comment (edgchen1) [inline tools/ci_build/build.py:1357]: thanks for moving these out! :)

File: orttraining/orttraining/test/python/launch_test.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+import logging
+
+logging.basicConfig(
+    format="%(asctime)s %(name)s [%(levelname)s] - %(message)s",
+    level=logging.DEBUG)
+
+log = logging.getLogger("Build")
+
+def run_subprocess(args, cwd=None, env={}):
```
Comment (edgchen1) [inline orttraining/orttraining/test/python/launch_test.py:18]: can this common function be refactored out and reused?

File: orttraining/orttraining/test/python/launch_test.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    completed_process = subprocess.run(args, cwd=cwd, check=True, env=my_env)
+    log.debug("Subprocess completed. Return code=" +
+              str(completed_process.returncode))
+    return completed_process
+
+def parse_arguments():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--process_args", required=True, help="process args.")
+    parser.add_argument("--cwd", help="cwd")
+    parser.add_argument("--dll_path", help="dll path.")
```
Comment (edgchen1) [inline orttraining/orttraining/test/python/launch_test.py:31]: is --dll_path used anywhere?

File: orttraining/orttraining/test/python/launch_test.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    my_env = os.environ.copy()
+    my_env.update(env)
+    completed_process = subprocess.run(args, cwd=cwd, check=True, env=my_env)
+    log.debug("Subprocess completed. Return code=" +
+              str(completed_process.returncode))
+    return completed_process
+
+def parse_arguments():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--process_args", required=True, help="process args.")
```
Comment (edgchen1) [inline orttraining/orttraining/test/python/launch_test.py:29]: help text could be more... helpful. for example, maybe it's worth mentioning that --process_args expects a single string containing the arguments

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-frontend-test-ci-pipeline.yml [hunk 5]
```diff
@@ -1,42 +1,257 @@
...
+        --volume $(Build.BinariesDirectory):/build \
+        --volume /bert_data:/bert_data \
+        onnxruntime-ubuntu16.04-cuda10.1-cudnn7.6 \
+          /build/RelWithDebInfo/launch_test.py \
+            --process_args "python orttraining_test_transformers.py BertModelTest.test_for_pretraining_mixed_precision -v" \
+            --cwd /build/RelWithDebInfo
+    displayName: 'Run orttraining_test_transformers.py BertModelTest.test_for_pretraining_mixed_precision'
+    condition: succeededOrFailed() # ensure all tests are run
+
+  # I have tried following to ways to run run_training_pipeline_e2e_tests. both failed.
```
Comment (xzhu1900) [inline tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-frontend-test-ci-pipeline.yml:225]: >run_training_pipeline_e2e_tests [](start = 42, length = 31)

What's the plan for pipelin_e2d_tests? Do you plan to move it together with run_training_python_frontend_e2e_tests and run_training_python_frontend_tests?

File: orttraining/orttraining/test/python/launch_test.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    parser.add_argument("--cwd", help="cwd")
+    parser.add_argument("--dll_path", help="dll path.")
+    parser.add_argument("--env", help="env variables.")
```
Comment (edgchen1) [inline orttraining/orttraining/test/python/launch_test.py:32]: how about:
```
parser.add_argument("--env", help="...", nargs=2, action="append", default=[])
```
to get a list of 2 element lists? can let argparse do the validation

File: orttraining/orttraining/test/python/launch_test.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+log = logging.getLogger("Build")
+
+def run_subprocess(args, cwd=None, env={}):
+    log.info("Running subprocess in '{0}'\n{1}".format(cwd or os.getcwd(), args))
+    my_env = os.environ.copy()
+    my_env.update(env)
+    completed_process = subprocess.run(args, cwd=cwd, check=True, env=my_env)
+    log.debug("Subprocess completed. Return code=" +
+              str(completed_process.returncode))
```
Comment (edgchen1) [inline orttraining/orttraining/test/python/launch_test.py:24]: given the call to subprocess.run() with check=True, won't return code always be 0 if it gets this far?

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-frontend-test-ci-pipeline.yml [hunk 8]
```diff
@@ -1,42 +1,257 @@
...
-      --enable_training_python_frontend_e2e_tests
-      --enable_training_pipeline_e2e_tests
-      "
-    displayName: 'Build and run frontend tests'
-
-  - template: templates/clean-agent-build-directory-step.yml
\ No newline at end of file
+  # update these if the E2E test data changes
+  - script: |
+      orttraining/tools/ci_test/download_e2e_test_data.py \
```
Comment (edgchen1) [inline tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-frontend-test-ci-pipeline.yml:15]: nit: can we rename this to download_azure_blob.py or something more general?

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-frontend-test-ci-pipeline.yml [hunk 9]
```diff
@@ -1,42 +1,257 @@
...
+        --volume /bert_data:/bert_data \
+        onnxruntime-ubuntu16.04-cuda10.1-cudnn7.6 \
+          /build/RelWithDebInfo/launch_test.py \
+            --process_args "python orttraining_test_transformers.py BertModelTest.test_for_pretraining_mixed_precision -v" \
+            --cwd /build/RelWithDebInfo
+    displayName: 'Run orttraining_test_transformers.py BertModelTest.test_for_pretraining_mixed_precision'
+    condition: succeededOrFailed() # ensure all tests are run
+
+  # I have tried following to ways to run run_training_pipeline_e2e_tests. both failed.
+  # For now, the --enable_training_pipeline_e2e_tests flag still triggles
```
Comment (edgchen1) [inline tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-frontend-test-ci-pipeline.yml:226]: "triggers"
just curious, what's the error?

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-frontend-test-ci-pipeline.yml [hunk 10]
```diff
@@ -1,42 +1,257 @@
...
+        --volume $(Build.BinariesDirectory)/training_e2e_test_data:/training_e2e_test_data:ro \
+        onnxruntime-ubuntu16.04-cuda10.1-cudnn7.6 \
+          /onnxruntime_src/orttraining/tools/ci_test/run_convergence_test.py \
+            --binary_dir /build/RelWithDebInfo \
+            --model_root /training_e2e_test_data/models \
+            --training_data_root /training_e2e_test_data/data
+    displayName: 'Run convergence test'
+    condition: succeededOrFailed() # ensure all tests are run
+
+    # migrated from frontend 2e2 pipeline
```
Comment (edgchen1) [inline tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-frontend-test-ci-pipeline.yml:69]: "e2e"

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-frontend-test-ci-pipeline.yml (hunk 5 above)
Comment (liqunfu) [reply tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-frontend-test-ci-pipeline.yml]: I am still working on it, however I tried 2 methods and both has failed. There are other issues with this PR. I may come to you for help once I am to fix it again.

File: orttraining/orttraining/test/python/_test_commons.py [hunk 11]
```diff
@@ -1,4 +1,47 @@
...
+import os
+import sys
+import subprocess
+
+def is_windows():
+    return sys.platform.startswith("win")
+
+def run_subprocess(args, cwd=None, capture=False, dll_path=None,
+                   shell=False, env={}, log=None):
+    log.info("Running subprocess in '{0}'\n{1}".format(
```
Comment (edgchen1) [inline orttraining/orttraining/test/python/_test_commons.py:11]: need test for log not being None?

File: orttraining/orttraining/test/python/_test_commons.py [hunk 12]
```diff
@@ -1,4 +1,47 @@
...
+    my_env.update(env)
+    completed_process = subprocess.run(
+        args, cwd=cwd, check=True, stdout=stdout, stderr=stderr,
+        env=my_env, shell=shell)
+    log.debug("Subprocess completed. Return code=" +
+              str(completed_process.returncode))
+    return completed_process
+
+
+def run_subprocess(args, cwd=None, env={}, log=None):
```
Comment (edgchen1) [inline orttraining/orttraining/test/python/_test_commons.py:34]: why do we need two run_subprocess() functions?

File: orttraining/orttraining/test/python/launch_test.py [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+
+print("sys.executable: ", sys.executable)
+cmd_line_with_args = launch_args.cmd_line_with_args.split()
+for n, arg in enumerate(cmd_line_with_args):
+    if arg == 'python':
+        cmd_line_with_args[n] = sys.executable
+
+cwd = launch_args.cwd
+
+env = dict(launch_args.env) if launch_args.env else {}
```
Comment (edgchen1) [inline orttraining/orttraining/test/python/launch_test.py:40]: how about `env = dict(launch_args.env)`?

File: orttraining/orttraining/test/python/_test_commons.py (hunk 12 above)
Comment (liqunfu) [reply orttraining/orttraining/test/python/_test_commons.py:34]: My mistake. I was using the shorter one then decided to use the full method from build.py. Forgot to remove the shorter one

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-e2e-test-nightly-pipeline.yml [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+          --enable_training \
+          --update --build \
+          --build_wheel \
+          --enable_training_python_frontend_e2e_tests \
+          --enable_training_pipeline_e2e_tests \
+          "
+    displayName: 'Build'
+
+  # Hit OOM with run_training_pipeline_e2e_tests.py - slightly above 16GB limit.
+  # leave this code here for further investigation.
```
Comment (xzhu1900) [inline tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-e2e-test-nightly-pipeline.yml:41]: Could you put the task link here for reference?

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-e2e-test-nightly-pipeline.yml [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+  # - script: |
+  #     docker run \
+  #       --gpus all \
+  #       --shm-size=256m \
```
Comment (xzhu1900) [inline tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-e2e-test-nightly-pipeline.yml:45]: >shm-size=256m [](start = 12, length = 13)

Any particular reason why this test's shared memory size is 256m while others are1024m?

File: orttraining/orttraining/test/python/launch_test.py (hunk 2 above)
Comment (liqunfu) [reply orttraining/orttraining/test/python/launch_test.py]: refactored to a common method called by multiple scripts.

File: orttraining/orttraining/test/python/launch_test.py (hunk 7 above)
Comment (liqunfu) [reply orttraining/orttraining/test/python/launch_test.py]: reverted to the original method with check as a optional input

File: orttraining/orttraining/test/python/launch_test.py (hunk 2 above)
Comment (liqunfu) [reply orttraining/orttraining/test/python/launch_test.py]: refactored into a common method called from multiple scripts

File: orttraining/orttraining/test/python/launch_test.py (hunk 3 above)
Comment (liqunfu) [reply orttraining/orttraining/test/python/launch_test.py]: removed as part of refactoring

File: orttraining/orttraining/test/python/launch_test.py (hunk 4 above)
Comment (liqunfu) [reply orttraining/orttraining/test/python/launch_test.py]: changed to cmd_line_with_args to be  clearer.  Extra help comments added

File: orttraining/orttraining/test/python/launch_test.py (hunk 7 above)
Comment (liqunfu) [reply orttraining/orttraining/test/python/launch_test.py]: reverted to the original subprocess

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-frontend-test-ci-pipeline.yml (hunk 8 above)
Comment (liqunfu) [reply tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-frontend-test-ci-pipeline.yml]: name changed

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-frontend-test-ci-pipeline.yml (hunk 9 above)
Comment (liqunfu) [reply tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-frontend-test-ci-pipeline.yml]: it was the missing of:
if __name__ == "__main__":
  sys.exit(main())

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-e2e-test-nightly-pipeline.yml (hunk 15 above)
Comment (liqunfu) [reply tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-e2e-test-nightly-pipeline.yml:45]: I was experimenting these numbers to see whether it make any difference. Moved back to 1024m.

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-e2e-test-nightly-pipeline.yml (hunk 14 above)
Comment (liqunfu) [reply tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-e2e-test-nightly-pipeline.yml:41]: added link

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-frontend-test-ci-pipeline.yml (hunk 5 above)
Comment (liqunfu) [reply tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-frontend-test-ci-pipeline.yml]: adding following code in teh script seemed to work. 
if __name__ == "__main__":
    sys.exit(main())

**Review by xzhu1900 — APPROVED** [review]
:shipit:

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-e2e-test-nightly-pipeline.yml [hunk 16]
```diff
@@ -0,0 +1,253 @@
...
+  #       --rm \
+  #       --volume $(Build.BinariesDirectory):/build \
+  #       --volume /bert_data:/bert_data \
+  #       --volume /bert_ort:/bert_ort \
+  #       onnxruntime-ubuntu16.04-cuda10.1-cudnn7.6 \
+  #         /build/RelWithDebInfo/run_training_pipeline_e2e_tests.py \
+  #           --cwd /build/RelWithDebInfo
+  #   displayName: 'Run run_training_pipeline_e2e_tests.py'
+  #   condition: succeededOrFailed() # ensure all tests are run
+  #   timeoutInMinutes: 10
```
Comment (thiagocrepaldi) [inline tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-e2e-test-nightly-pipeline.yml:56]:
```suggestion
```

**Review by thiagocrepaldi — APPROVED** [review]

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-e2e-test-nightly-pipeline.yml (hunk 16 above)
Comment (liqunfu) [reply tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-e2e-test-nightly-pipeline.yml:56]: Discussed with Xueyun. She is to make run_training_pipeline_e2e_tests.py pass and uncomment this block after merge of this PR.

**Pull request merged**